native-maven-plugin/src/main/java/org/graalvm/buildtools/utils/AgentUtils.java

```diff
@@ -0,0 +1,207 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.graalvm.buildtools.utils.Utils.parseBoolean;
+
+public final class AgentUtils {
+
+    private AgentUtils() { }
+
+    public static AgentMode getAgentMode(Xpp3Dom agent) throws Exception {
+        Xpp3Dom defaultModeNode = Xpp3DomParser.getTagByName(agent, "defaultMode");
+        if (defaultModeNode == null) {
+            throw new RuntimeException("Default agent mode not provided in configuration.");
+        }
+
+        Xpp3Dom agentModes = Xpp3DomParser.getTagByName(agent, "modes");
+
+        AgentMode agentMode;
+        String mode = defaultModeNode.getValue();
+        switch (mode.toLowerCase()) {
+            case "standard":
+                agentMode = new StandardAgentMode();
+                break;
+            case "disabled":
+                agentMode = new DisabledAgentMode();
+                break;
+            case "conditional":
+                // conditional mode needs few more options declared in xml
+                if (agentModes == null) {
+                    throw new RuntimeException("AgentModes tag not provided in configuration.");
+                }
+
+                Xpp3Dom userCodeFilterPathNode = Xpp3DomParser.getTagByName(agentModes, "userCodeFilterPath");
+                Xpp3Dom extraFilterPathNode = Xpp3DomParser.getTagByName(agentModes, "extraFilterPath");
+
+                if (userCodeFilterPathNode == null) {
+                    throw new Exception("UserCodeFilterPath must be provided in agent configuration");
+                }
+
+                agentMode = new ConditionalAgentMode(userCodeFilterPathNode.getValue(),
+                                                    extraFilterPathNode != null ? extraFilterPathNode.getValue() : "",
+                                                     parseBooleanNode(agentModes, "parallel"));
+                break;
+            case "direct":
+                // direct mode is given
+                if (agentModes == null) {
+                    throw new RuntimeException("AgentModes tag not provided in configuration.");
+                }
+
+                Xpp3Dom directModeNode = Xpp3DomParser.getTagByName(agentModes, "direct");
+                if (directModeNode == null) {
+                    throw new RuntimeException("Direct agent mode not provided in configuration.");
+                }
+
+                List<String> options = Arrays.stream(directModeNode.getValue().split(" ")).collect(Collectors.toList());
+                agentMode = new DirectAgentMode(options);
+                break;
+            default:
+                throw new Exception("Unknown agent mode selected: " + mode);
+        }
+
+        return agentMode;
+    }
+
+    public static AgentConfiguration collectAgentProperties(MavenSession session, Xpp3Dom rootNode) throws RuntimeException {
```

Comment:
I thought one of the goal of the refactoring was to stop doing manual parsing of the nodes and leverage regular binding to also benefit from autocomplete in the IDE. Could you please elaborate on the limit that prevent us to use regular binding and if if we now have a good DevXP (no "everything in red" when we configure the agent and full auto-complete in the IDE)?

Reply:
Since we are using agent configuration in the `NativeExtension` as well, we need to parse pom file manually, to collect configuration in this stage. Also we are using regular binding as well, to prevent "everything in red" and to use agent options in two mojos.